src/core/server/environment/write_pid_file.ts

```diff
@@ -0,0 +1,68 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { unlinkSync as unlink } from 'fs';
+import once from 'lodash/once';
+import { Logger } from '../logging';
+import { writeFile, exists } from './fs';
+import { PidConfigType } from './pid_config';
+
+export const writePidFile = async ({
+  pidConfig,
+  logger,
+}: {
+  pidConfig: PidConfigType;
+  logger: Logger;
+}) => {
+  const path = pidConfig.file;
+  if (!path) {
+    return;
+  }
+
+  const pid = String(process.pid);
+
+  if (await exists(path)) {
+    const message = `pid file already exists at ${path}`;
+    if (pidConfig.exclusive) {
+      throw new Error(message);
+    } else {
+      logger.warn(message, { path, pid });
+    }
+  }
+
+  await writeFile(path, pid);
+
+  logger.info(`wrote pid file to ${path}`, { path, pid });
+
+  const clean = once(() => {
+    unlink(path);
+  });
+
+  process.once('exit', clean); // for "natural" exits
+  process.once('SIGINT', () => {
+    // for Ctrl-C exits
+    clean();
+    // resend SIGINT
+    process.kill(process.pid, 'SIGINT');
+  });
+
+  process.on('unhandledRejection', function (reason) {
+    logger.warn(`Detected an unhandled Promise rejection.\n${reason}`);
+  });
```

Comment:
This was in the legacy file, so I kept it for now, but I honestly have no idea why this was here. 

This seems already handled by https://github.com/elastic/kibana/blob/bf04235dae35452061cc7ea3d86d96c19a58206c/src/setup_node_env/exit_on_warning.js#L39-L45

@spalger Am I safe to remove this from this file?

Reply:
I wonder if this is here so that warnings are included in the standard kibana log, rather than only written to the console, once the logger is setup. That said, both of these listeners do seem to be setup. I'd ask @watson just to be sure though.

(sorry for the late handoff, was out last week)

Reply:
I must have missed the original `unhandledRejection` listener from `src/legacy/server/pid/index.js` when I implemented the one in `src/setup_node_env/exit_on_warning.js`.

The intention with the `src/setup_node_env/exit_on_warning.js` script is to exit on warnings during development and CI only. This way we catch these warnings as early as possible. It's all too easy to overlook them in the logs during development/CI, and so they slip into production builds. In production we normally run with the `--no-warnings` flag, so none of the listeners in `src/setup_node_env/exit_on_warning.js` are added in production.

So in production, the listener in this file is still active (though output depends on log-level). Whether or not we want this in production is a good question that I don't know the answer to. It's worth noting that Node.js it self will output the following to STDERR in case there's no `unhandledRejection` listener:

> (node:88082) UnhandledPromiseRejectionWarning: undefined
> (node:88082) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 1)
> (node:88082) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.

So adding the listener in this file means that we'll suppress that default behaviour even if log-level is set below `warning`.

_Side note: Starting in Node.js 15, the process will crash by default if an `unhandledRejection` happens, but we'll not be affected by this until we upgrade to Node.js 16 as we'll skip 15._

Reply:
Thanks for the answer. To avoid changing the behavior that was present in legacy, I will keep this now.